charmhelpers/contrib/openstack/audits/__init__.py

```
@@ -0,0 +1,140 @@
+
+
+_audits = {}
+
+Audit = collections.namedtuple('Audit', 'func filters')
+
+
+def audit(*args):
+    """Decorator to register an audit.
+
+    These are used to generate audits that can be run on a
+    deployed system that matches the given configuration
+
+    :param args: List of functions to filter tests against
+    :type args: List[Callable(Config)]
+    """
+    def wrapper(f):
+        test_name = f.__name__
+        if _audits.get(test_name):
+            raise RuntimeError(
+                "Test name '{}' used more than once ?"
+                .format(test_name))
+        _audits[test_name] = Audit(func=f, filters=args)
+        return f
+    return wrapper
+
+
+def is_audit_type(*args):
+    """This audit is included in the specified kinds of audits."""
+    def should_run(config):
+        if config.get('audit_type') in args:
+            return True
+        else:
+            return False
+    return should_run
+
+
+def since_package(pkg, pkg_version):
+    """This audit should be run after the specified package version (incl)."""
+    return lambda config=None: not before_package(pkg, pkg_version)()
+
+
+def before_package(pkg, pkg_version):
+    """This audit should be run before the specified package version (excl)."""
+    def should_run(config=None):
+        version = get_upstream_version(pkg)
+        return version and \
+            (apt_pkg.version_compare(version, pkg_version) < 0)
+    return should_run
+
+
+def it_has_config(config_key):
+    """This audit should be run based on specified config keys."""
+    return lambda config: config.get(config_key) is not None
+
+
+def run(config):
+    """Run the configured audits with the specified config.
+
+    :param config: Configuration for the audit
+    :type config: Config
+    """
+    errors = {}
+    results = {}
+    for name, audit in sorted(_audits.items()):
+        result_name = name.replace('_', '-')
+        if all(p(config) for p in audit.filters):
```

It would probably be sensible to have a test that `p` is `callable` before calling it.  That way you can pick up any non callables before they get to here; it's probably best to put a check in the decorator that the arguments are all callable that let it fail here?